### PR TITLE
Move List (the type) into metav1 but preserve the exposed type

### PR DESCRIPTION
**Pull request by smarterclayton**
Make a list something that other components can use without depending on the core API. This does not change the public API exposure of `List` (it is still in `v1`) but makes the interface common across both.

### CODE REVIEW

File: staging/src/k8s.io/api/admission/v1alpha1/types.generated.go [hunk 1]
```diff
@@ -1141,13 +1141,7 @@ func (x *AdmissionReviewStatus) CodecEncodeSelf(e *codec1978.Encoder) {
 					if x.Result == nil {
 						r.EncodeNil()
 					} else {
-						yym7 := z.EncBinary()
```
Comment (sttts) [inline staging/src/k8s.io/api/admission/v1alpha1/types.generated.go]: these changes look wrong.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/doc.go [hunk 2]
```diff
@@ -1,19 +0,0 @@
-/*
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/doc.go]: why delete this?

File: pkg/api/types.go [hunk 3]
```diff
@@ -3593,13 +3593,7 @@ type EventList struct {
...
 
 // List holds a list of objects, which may not be known by the server.
-type List struct {
-	metav1.TypeMeta
-	// +optional
-	metav1.ListMeta
-
-	Items []runtime.Object
-}
+type List metainternalversion.List
```
Comment (sttts) [inline pkg/api/types.go]: How about deprecating this?

Comment (sttts) [inline pkg/api/types.go]: I mean we could register the meta List type in core/v1, but deprecate it on the code level.

**Comment by sttts** [comment]
In general 👍  The many ugorji changes look wrong.

**Comment by deads2k** [comment]
I'm in favor of the move.

File: pkg/api/types.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/api/types.go]: Hrm, good point.  We could probably also remove this alias.  Let me look at how broad that is.

**Comment by smarterclayton** [comment]
Somethings wrong with ugorji generation after I make this move.  Going to have to dig in more to find out why

File: pkg/api/types.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/api/types.go]: Looked into it, something about deep copy generation isn't right in metainternalversion.  I'll do this in a follow up - sound ok?

Comment (sttts) [inline pkg/api/types.go]: Sure. Actually I am surprised that this alias works, compare https://github.com/kubernetes/gengo/pull/73.

Comment (smarterclayton) [inline pkg/api/types.go]: I had to add alias resolution a while back for protobuf, so some parts of aliasing do work.

**Comment by smarterclayton** [comment]
All fixed now.  Any other comments beyond the followup?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 4]
```diff
@@ -679,6 +680,18 @@ const (
 	CauseTypeUnexpectedServerResponse CauseType = "UnexpectedServerResponse"
 )
 
+// List holds a list of objects, which may not be known by the server.
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go [hunk 5]
```diff
@@ -48,3 +49,12 @@ type ListOptions struct {
 	// Timeout for the list/watch call.
 	TimeoutSeconds *int64
 }
+
+// List holds a list of objects, which may not be known by the server.
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go]: // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object

**Comment by sttts** [comment]
staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/doc.go must not be dropped. It also drops the `deepcopy=package` tag and now deepcopy is not regenerated there.

**Comment by smarterclayton** [comment]
Updated and green.

**Pull request merged**